src/Microsoft.Management.Configuration/ConfigurationSetSerializer.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#include "pch.h"
+
+#include <AppInstallerLogging.h>
+#include <AppInstallerVersions.h>
+#include <AppInstallerStrings.h>
+
+#include "ConfigurationSetSerializer.h"
+#include "ConfigurationSetSerializer_0_2.h"
+#include "ConfigurationSetUtilities.h"
+
+using namespace AppInstaller::YAML;
+using namespace winrt::Windows::Foundation;
+
+namespace winrt::Microsoft::Management::Configuration::implementation
+{
+    std::unique_ptr<ConfigurationSetSerializer> ConfigurationSetSerializer::CreateSerializer(hstring version)
+    {
+        // Create the parser based on the version selected
+        AppInstaller::Utility::SemanticVersion schemaVersion(std::move(winrt::to_string(version)));
+
+        // TODO: Consider having the version/uri/type information all together in the future
+        if (schemaVersion.PartAt(0).Integer == 0 && schemaVersion.PartAt(1).Integer == 1)
+        {
+            throw E_NOTIMPL;
+        }
+        else if (schemaVersion.PartAt(0).Integer == 0 && schemaVersion.PartAt(1).Integer == 2)
+        {
+            return std::make_unique<ConfigurationSetSerializer_0_2>();
+        }
+        else if (schemaVersion.PartAt(0).Integer == 0 && schemaVersion.PartAt(1).Integer == 3)
+        {
+            throw E_NOTIMPL;
+        }
+        else
+        {
+            AICLI_LOG(Config, Error, << "Unknown configuration version: " << schemaVersion.ToString());
+            throw E_UNEXPECTED;
+        }
+    }
+
+    void ConfigurationSetSerializer::WriteYamlValueSet(AppInstaller::YAML::Emitter& emitter, const Windows::Foundation::Collections::ValueSet& valueSet)
+    {
+        emitter << BeginMap;
+
+        for (const auto& [key, value] : valueSet)
+        {
+            IPropertyValue property = value.try_as<IPropertyValue>();

```

Comment:
The values in a `ValueSet` can be both `IPropertyValue` and `ValueSet`.  This code needs to handle the `ValueSet` case as well.